### ISSUE TITLE
Wrong reflection metadata on subclasses

### ISSUE THREAD

**Issue by mhalbritter** [oracle/graal#4809]
```
openjdk version "17.0.4" 2022-07-19
OpenJDK Runtime Environment GraalVM CE 22.2.0 (build 17.0.4+8-jvmci-22.2-b06)
OpenJDK 64-Bit Server VM GraalVM CE 22.2.0 (build 17.0.4+8-jvmci-22.2-b06, mixed mode, sharing)
```

Hey, we're currently having problems with a smoke test of ours. We tried to create a more minimal example but failed. The smoke test uses netty, reactor netty, and lettuce.

The problem is that when running in native-image, netty drops all messages received. We tracked this down to Netty's `ChannelHandlerMask`, which uses reflection to check if a method is skippable, signalled by the `@Skip` annotation. All methods from the `ChannelInboundHandlerAdapter`, which is often used to implement a custom channel handler, are skippable.

The main problem lies in `reactor.netty.transport.ServerTransport.Acceptor`, which extends `ChannelInboundHandlerAdapter` and overrides two method without applying the `@Skip` annotation: `channelRead` and `exceptionCaught`.

When running on the JVM, these methods are correctly marked as not skippable, as seen in this logs:

```
2022-08-11T15:15:29.951+02:00  INFO 431335 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.channelRead([interface io.netty.channel.ChannelHandlerContext, class java.lang.Object])
2022-08-11T15:15:29.951+02:00  INFO 431335 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T15:15:29.951+02:00  INFO 431335 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.exceptionCaught([interface io.netty.channel.ChannelHandlerContext, class java.lang.Throwable])
2022-08-11T15:15:29.951+02:00  INFO 431335 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T15:15:29.951+02:00  INFO 431335 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Mask of class reactor.netty.transport.ServerTransport$Acceptor: 33
```

which results in a mask of 33. I've added our own copy of `ChannelHandlerMask` with debug logging in there to get these logs.

Now things get weird. When running with lettuce on the classpath and no metadata on `ServerTransport$Acceptor` in native-image:

```
2022-08-11T14:18:44.871+02:00  INFO 421297 --- [           main] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.channelRead([interface io.netty.channel.ChannelHandlerContext, class java.lang.Object])
2022-08-11T14:18:44.871+02:00  INFO 421297 --- [           main] io.netty.channel.ChannelHandlerMask      : Method found, skippable: true
2022-08-11T14:18:44.871+02:00  INFO 421297 --- [           main] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.exceptionCaught([interface io.netty.channel.ChannelHandlerContext, class java.lang.Throwable])
2022-08-11T14:18:44.871+02:00  INFO 421297 --- [           main] io.netty.channel.ChannelHandlerMask      : Method found, skippable: true
2022-08-11T14:18:44.871+02:00  INFO 421297 --- [           main] io.netty.channel.ChannelHandlerMask      : Mask of class reactor.netty.transport.ServerTransport$Acceptor: 0
```

Now Netty thinks that these methods are skippable, skips them and drops all messages. This is what's causing the test to fail. This can only happen if `m.isAnnotationPresent(Skip.class)` in `ChannelHandlerMask.isSkippable` returns true. The methods in `ServerTransport$Acceptor` are NOT annotated with `@Skip` and `@Skip` is not marked as `@Inherited`, so there is clearly something wrong here.

We can fix this by adding reflection metadata for `reactor.netty.transport.ServerTransport$Acceptor` (which is what ` @ImportRuntimeHints({ NettyRuntimeHints.class, ReactorNettyHints.class })` does):

When running with lettuce on the classpath and hints on `ServerTransport$Acceptor`:

```
2022-08-11T14:21:57.763+02:00  INFO 422057 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.channelRead([interface io.netty.channel.ChannelHandlerContext, class java.lang.Object])
2022-08-11T14:21:57.763+02:00  INFO 422057 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T14:21:57.763+02:00  INFO 422057 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.exceptionCaught([interface io.netty.channel.ChannelHandlerContext, class java.lang.Throwable])
2022-08-11T14:21:57.763+02:00  INFO 422057 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T14:21:57.763+02:00  INFO 422057 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Mask of class reactor.netty.transport.ServerTransport$Acceptor: 33
```

Now these methods marked correctly as not skippable and the mask is 33, which is the same as on the JVM (verify with `gradle bootRun`) and everything works again.

Maybe you wondered why i mentioned lettuce. Including the `io.lettuce:lettuce-core` dependency has an effect, because this JAR contains reflection hints on [classes which extend ChannelInboundAdapter](https://github.com/lettuce-io/lettuce-core/blob/main/src/main/resources/META-INF/native-image/io.lettuce/lettuce-core/reflect-config.json#L8).

This, to my surprise, causes reflection access on other subclasses of `ChannelInboundAdapter` to succeed, like the `ServerTransport$Acceptor`. And I guess there is the problem somewhere: By adding hints on a lettuce subclass of `ChannelInboundAdapter` some reflection metadata
is generated for `ChannelInboundAdapter`, but this reflection metadata is wrong for other subclasses of `ChannelInboundAdapter`, like the `ServerTransport$Acceptor`.

When running without lettuce metadata on the classpath and no hints on `ServerTransport$Acceptor`:

```
2022-08-11T15:08:34.162+02:00  INFO 430555 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.channelRead([interface io.netty.channel.ChannelHandlerContext, class java.lang.Object])
2022-08-11T15:08:34.162+02:00  INFO 430555 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method not found, assuming not skippable
2022-08-11T15:08:34.162+02:00  INFO 430555 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.exceptionCaught([interface io.netty.channel.ChannelHandlerContext, class java.lang.Throwable])
2022-08-11T15:08:34.162+02:00  INFO 430555 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method not found, assuming not skippable
2022-08-11T15:08:34.162+02:00  INFO 430555 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Mask of class reactor.netty.transport.ServerTransport$Acceptor: 511
```

Now the reflection fails on these methods, Netty assumes they are not skippable (that's why the mask has changed).

When running without lettuce metadata on the classpath and hints on `ServerTransport$Acceptor`:

```
2022-08-11T15:11:28.727+02:00  INFO 430991 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.channelRead([interface io.netty.channel.ChannelHandlerContext, class java.lang.Object])
2022-08-11T15:11:28.727+02:00  INFO 430991 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T15:11:28.727+02:00  INFO 430991 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Inspecting class reactor.netty.transport.ServerTransport$Acceptor.exceptionCaught([interface io.netty.channel.ChannelHandlerContext, class java.lang.Throwable])
2022-08-11T15:11:28.727+02:00  INFO 430991 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Method found, skippable: false
2022-08-11T15:11:28.727+02:00  INFO 430991 --- [ctor-http-nio-1] io.netty.channel.ChannelHandlerMask      : Mask of class reactor.netty.transport.ServerTransport$Acceptor: 33
```

which is correct again.

## How to reproduce

I've created a fork of the smoke tests with the debugs logging. It's named `session-redis-webflux` and can be found [here](https://github.com/mhalbritter/spring-aot-smoke-tests/tree/main/session-redis-webflux). I've added comments how to include / exclude hints for the `ServerTransport$Acceptor`.

To run it, execute `gradle nativeRun` and see the startup logs. You need a running redis, too: `docker run -p 6379:6379 redis:latest`

To get a lettuce-core without hints, the easiest possibility is to clone lettuce-core [from here](https://github.com/lettuce-io/lettuce-core), check out the `6.2.0.RELEASE` tag, change the `version` in the `pom.xml` to `6.2.0.RELEASE-NO-METADATA`, remove the files under `src/main/resources/META-INF/native-image` and run a `mvn clean install`. Then comment in the `resolutionStrategy` line in `build.gradle` in the smoke test.

Thanks and please label this as a `spring` issue.

**Comment by oubidar-Abderrahim** [comment]
Hi, Thank you for reporting this, I see that the issue created for reactor-netty was closed, do the new changes fix the issue?

**Comment by violetagg** [comment]
@oubidar-Abderrahim The issue in Reactor Netty is not an issue but enhancement. It does not fix anything related to the issue described here.

**Comment by loicottet** [comment]
This is caused by what we call "hiding methods", which we use to mark methods that are not registered for reflection but override methods that are registered. Without this, calling `getMethod` on the class that holds the overriding method would return the superclass method.

In this case, we are mistakenly not marking `ServerTransport$Acceptor.channelRead` as hiding due to the following events happening:

1. `ChannelInboundHandlerAdapter.channelRead` is registered for reflection. `ServerTransport$Acceptor` is not yet seen as reachable by the analysis.
2. The reachable subclasses of `ChannelInboundHandlerAdapter` are searched for an overriding method. Since `ServerTransport$Acceptor` is not yet reachable, `ServerTransport$Acceptor.channelRead` is not marked as hiding.
3. `ServerTransport$Acceptor` is then marked as reachable by the analysis.
4. `ChannelHandlerMask.isSkippable` is called on `ServerTransport$Acceptor.channelRead` at runtime, and returns the wrong method since Native Image believes `ServerTransport$Acceptor` does not override `channelRead`, as it was neither registered for reflection nor marked as hiding.

I believe the hints in `io.lettuce:lettuce-core` result in `ServerTransport$Acceptor` being seen as reachable before `ChannelInboundHandlerAdapter.channelRead` is registered for reflection, which means `ServerTransport$Acceptor.channelRead` is then correctly marked as hiding.

The way to solve this is to look for hiding methods during each analysis iteration, instead of only when registering the parent method. I will create a fix for this issue.

**Comment by loicottet** [comment]
This should have been fixed by #4835. Can you please confirm it is the case ?

**Comment by asantare** [comment]
hi @loicottet, using the 22.3-dev (where I guess/hope these changes are present), it worked fine for the linked netty issue, leaving to mhalbritter the test of this one

That you know, will the changes be present in a 22.2 bugfix? or only in the 22.3?

**Comment by loicottet** [comment]
The fix will indeed be present in 22.3, there is no plan to backport it to 22.2.

**Comment by sdeleuze** [comment]
@loicottet Could you please update the milestone of this issue accordingly?
